File: solutions/javascript/bfs/numberOfConnectedComponentsInAnUndirectedGraph.js
const buildGraph = (edges) => {
    const graph = {};

    for (const edge of edges) {
        const [nodeOne, nodeTwo] = edge;

        const nodeOneInGraph = graph.hasOwnProperty(nodeOne);
        const nodeTwoInGraph = graph.hasOwnProperty(nodeTwo);

        if (!nodeOneInGraph) graph[nodeOne] = [];
        if (!nodeTwoInGraph) graph[nodeTwo] = [];

        graph[nodeOne].push(nodeTwo);
        graph[nodeTwo].push(nodeOne);
    }

    return graph;
};

const markGroupAsVisited = (startNode, graph, visited) => {
    const queue = new Queue();
    queue.enqueue(startNode);

    while (queue.size() > 0) {
        // Remove node
        const node = queue.dequeue();

        // Process node
        visited.add(node);

        // Add neighbors
        const nodeInGraph = graph.hasOwnProperty(node);
        if (!nodeInGraph) continue;

        const neighbors = graph[node];

        for (const neighbor of neighbors) {
            if (visited.has(neighbor)) continue;

            queue.enqueue(neighbor);
        }
    }
};

const countComponents = (numNodes, edges) => {
    const graph = buildGraph(edges);

    let numConnectedComponents = 0;
    const visited = new Set();

    for (let node = 0; node < numNodes; node++) {
        if (visited.has(node)) continue;

        numConnectedComponents++;
        markGroupAsVisited(node, graph, visited);
    }

    return numConnectedComponents;
};

File: solutions/python/bfs/number_of_connected_components_in_an_undirected_graph.py
class Solution:
    def build_graph(self, edges):
        graph = defaultdict(list)
        
        for node_one, node_two in edges:
            graph[node_one].append(node_two)
            graph[node_two].append(node_one)
        
        return graph
    
    def mark_component_as_visited(self, graph, start_node, visited):
        queue = deque()
        queue.append(start_node)
        
        while queue:
            # Remove node
            node = queue.popleft()
            
            # Process node
            # No work required here for this problem
            
            # Add neighbors
            neighbors = graph[node]
            
            for neighbor in neighbors:
                if neighbor in visited:
                    continue
                visited.add(neighbor)
                queue.append(neighbor)
    
    def countComponents(self, num_nodes: int, edges: List[List[int]]) -> int:
        graph = self.build_graph(edges)
        
        num_connected_components = 0
        visited = set()
        
        for node in range(num_nodes):
            if node in visited:
                continue
                
            num_connected_components += 1
            self.mark_component_as_visited(graph, node, visited)
            
        return num_connected_components